To merge/libretext jwt

Miniland1333 opened the pull request:
Hello! I'm been working with my staging deployment and I have some small suggested modification to various env variables. There's likely a lot to discuss here so let me know what you think in a PR review.

Changes (in order of importance)

1. Added SSL libraries to Dockefile for HTTPS requests out.
2. Used `//=` operator with env variables so that config defaults are used if env are null/undefined. Note that `||=` is not acceptable.
3. Renamed `baseURL` and `formURL`.
4. Renamed `libraryURL` to `problemSourceURL`.
5. Added `max_redirects(5)` to `fetchRemoteSource_p`.
6. Renamed `form_data` to `inputs_Ref`.
7. Added `$SITE_HOST` to the rest of the templates.

Question: I've mentioned this before, but for a site with a baseURL = `/webwork` do you think that the SITE_HOST should also end with `/webwork`? This decision affects point (7) as I may need to substitute both SITE_HOST and baseURL into the templates instead of what we currently have.

drdrew42 commented on lib/WeBWorK/lib/WebworkClient/classic_format.pl:
```diff
@@ -4,33 +4,33 @@
 <meta charset='utf-8'>
 <base href="$SITE_URL">
-<link rel="shortcut icon" href="/webwork2_files/images/favicon.ico"/>
+<link rel="shortcut icon" href="$SITE_HOST/webwork2_files/images/favicon.ico"/>
```
Here, base href is `$SITE_URL` which is just an alias for `$ENV{baseURL}` which is used to construct `$webwork_htdocs_url` ...so this should be based off of `$webwork_htdocs_url` instead of `$SITE_HOST`

The base href doesn't work unless the renderedHTML is loaded into its own DOM document (own page or iframe), and generally, I think we should favor absolute URLs over relative ones... 

We have now: `baseURL` `SITE_HOST` and `formURL` -- they should be crafted in this way:
- `SITE_HOST` -- absolute URL, root endpoint of host (e.g. `https://myhost.com`)
- `baseURL` -- relative to `SITE_HOST`, extends to renderer root endpoint (e.g. `/webwork` as in your dev)
- `formURL` -- relative to `baseURL`, the renderer endpoint for problem submission (e.g. `/render-api`)

We also need to distinguish between these values as set in config, and their subsequent implementation as ENV variables. I think we construct `$ENV{baseURL}` as an absolute URL by concatenating with `SITE_HOST`. And then if any request overrides this, it is expected that the overriding value is an absolute URL. We should do the same for `$ENV{formURL}` -- which is where I can see more relevant use for overriding anyway.

Then, in these templates, use `$webwork_htdocs_url` instead of `$SITE_HOST/webwork2_files` -- that way the config file is the future entry point for making adjustments rather than having to modify each template.

Miniland1333 replied:
I believe this has now been accomplished for the renderer, though at a later time I need to add `$webwork_htdocs_url` to your Mojolicious application since it currently has hardcoded `/webwork2_files` in it.

drdrew42 reviewed (changes requested):
Overall, this looks great!

I'd like to see some cleanup with regards to our three URL configurations, see comments on `classic_format.pl`

Miniland1333 commented on lib/RenderApp.pm:
```diff
@@ -32,18 +32,31 @@ use RenderApp::Model::Problem;
+		$ENV{baseURL} = '';
+	}
+	#TODO
+	$ENV{baseURL} = $ENV{SITE_HOST}.$ENV{baseURL};
+	$ENV{formURL} = $ENV{baseURL}.$ENV{formURL};
```
I don't know how to do this in Perl, but there should be a check like this:
```javascript
if (!$ENV{baseURL}.startsWith('http'))
     $ENV{baseURL} = $ENV{SITE_HOST}.$ENV{baseURL};
if (!$ENV{formURL}.startsWith('http'))
     $ENV{formURL} = $ENV{baseURL}.$ENV{formURL};
```

This is so that if baseURL and formURL are passed in as relative paths, this then constructs the absolute path.

drdrew42 replied:
```perl
if ( $ENV{baseURL} !~ m|^https?://| ) {
     $ENV{baseURL} = $ENV{SITE_HOST}.$ENV{baseURL};
}
```
perl uses `=~` /regex/ and its negation `!~`; it also allows for alternate regex delimiters if prefaced with `m`. I have used them so that I don't need to escape the literal `/`s in the regex pattern.

Miniland1333 replied:
Thanks! I've implemented and tested those. Also, I've added a check to strip a trailing slash from SITE_HOST if present.

Miniland1333 commented:
Last changes I've made in these final commits:

- Added Dockerfile_with_OPL so that the OPL is bundled in the Dockerfile. Takes longer to build but is very portable.
- Added $problemSourceURL to template files so that you are able to submit/preview those problems.

That should be everything for this PR!
